src/kemal/middleware/filters.cr

```diff
@@ -2,47 +2,72 @@ module Kemal::Middleware
   # Kemal::Filter handle all code that should be evaluated before and after
   # every request
   class Filter < HTTP::Handler
+    INSTANCE = new
+
+    # This middleware is lazily instantiated and added to the handlers as soon as a call to `after_X` or `before_X` is made.
     def initialize
       @tree = Radix::Tree.new
+      Kemal.config.add_handler(self)
     end
 
-    def add(type, path, options, &block : HTTP::Server::Context -> _)
-      node = radix_path type, path
-      @tree.add node, Block.new &block
-    end
-
+    # The call order of the filters is before_all -> before_x -> X -> after_x -> after_all
     def call(context)
-      process_filter(context, :before)
+      return call_next(context) unless context.route_defined?
+      call_block_for_path_type("ALL", context.request.path, :before, context)
+      call_block_for_path_type(context.request.override_method, context.request.path, :before, context)
       call_next(context)
-      process_filter(context, :after)
+      call_block_for_path_type(context.request.override_method, context.request.path, :after, context)
+      call_block_for_path_type("ALL", context.request.path, :after, context)
+      context
     end
 
-    def filter_for_path_type_defined?(path, type)
-      lookup = @tree.find radix_path(type, path)
+    # This checks is filter is already defined for the verb/path/type combination
+    def filter_for_path_type_defined?(verb, path, type)
+      lookup = @tree.find radix_path(verb, path, type)
       lookup.found? && lookup.payload.is_a? Block
     end
 
-    private def process_filter(context, type)
-      lookup = @tree.find radix_path(type, context.request.path)
+    # This shouldn't be called directly, it's not private because I need to call it for testing purpose since I can't call the macros in the spec.
+    # It adds the block for the corresponding verb/path/type combination to the tree.
+    def add_route_filter(verb, path, type, &block : HTTP::Server::Context -> _)
```

Comment:
Since it is not public API, but it can't be called from user's perspective, it makes sense to put `# :nodoc:` comment, so that it does not show up in a documentation.

Additionally, I like to name such methods with leading `_`, but that is just personal preference/opinion.